plonk-core/src/lookup/lookup_table.rs

```diff
@@ -0,0 +1,368 @@
+        let upper_bound = 2u64.pow(n.into());
+        let range = lower_bound..upper_bound;
+        for a in range.clone() {
+            range
+                .clone()
+                .for_each(|b| self.insert_mul_row(a, b, upper_bound));
+        }
+    }
+
+    /// Function builds a table from mutiple operations. If, for example,
+    /// we are using lookup tables for both XOR and mul operataions, we can
+    /// create a table where the rows 0..n/2 use a mul function on all 2^n
+    /// indices and have the 4th wire storing index 0. For all indices n/2..n,
+    /// an XOR gate can be added, wheren the index of the 4th wire is 0.
+    /// These numbers require exponentiation outside, for the lower bound,
+    /// otherwise the range cannot start from zero, as 2^0 = 1.
+    /// Particular XOR row(s) can be added with this function.
+    pub fn insert_multi_xor(&mut self, lower_bound: u64, n: u8) {
+        let upper_bound = 2u64.pow(n.into());
+        let range = lower_bound..upper_bound;
+        for a in range.clone() {
+            range
+                .clone()
+                .for_each(|b| self.insert_xor_row(a, b, upper_bound));
+        }
+    }
+
+    /// Function builds a table from mutiple operations. If, for example,
+    /// we are using lookup tables for both XOR and mul operataions, we can
+    /// create a table where the rows 0..n/2 use a mul function on all 2^n
+    /// indices and have the 4th wire storing index 0. For all indices n/2..n,
+    /// an XOR gate can be added, wheren the index of the 4th wire is 0.
+    /// These numbers require exponentiation outside, for the lower bound,
+    /// otherwise the range cannot start from zero, as 2^0 = 1.
+    /// Particular AND row(s) can be added with this function.
+    pub fn insert_multi_and(&mut self, lower_bound: u64, n: u8) {
+        let upper_bound = 2u64.pow(n.into());
+        let range = lower_bound..upper_bound;
+        for a in range.clone() {
+            range
+                .clone()
+                .for_each(|b| self.insert_and_row(a, b, upper_bound));
+        }
+    }
+
+    /// Takes in a table, which is a vector of slices containing
+    /// 4 elements, and turns them into 4 distinct multisets for
+    /// a, b, c and d.
+    pub fn vec_to_multiset(&self) -> Vec<MultiSet<F>> {
+        let mut result = vec![MultiSet::new(); 4];
+        self.0.iter().for_each(|row| {
+            result.iter_mut().enumerate().for_each(|(index, multiset)| {
+                multiset.push(row[index]);
+            })
+        });
+        result
+    }
+
+    /// Attempts to find an output value, given two input values, by querying
+    /// the lookup table. The final wire holds the index of the table. The
+    /// element must be predetermined to be between -1 and 2 depending on
+    /// the type of table used. If the element does not exist, it will
+    /// return an error.
+    pub fn lookup(&self, a: F, b: F, d: F) -> Result<F, Error> {
```

Comment:
You should also be able to lookup over any subset of the relation arguments. These tables are not going to just be used for function tables.

Reply:
Agreed. This will be resolved as a part of #131